sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -404,82 +354,69 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
+    } else {
+      parsed
     }
-  }
-
-  private def getTable3LNamespace(tableName: String): Table = {
-    val ident = sparkSession.sessionState.sqlParser.parseMultipartIdentifier(tableName)
-    makeTable(ident)
+    makeTable(nameParts)
   }
 
   /**
    * Gets the table or view with the specified name in the specified database. This throws an
    * `AnalysisException` when no `Table` can be found.
    */
   override def getTable(dbName: String, tableName: String): Table = {
-    if (tableExists(dbName, tableName)) {
-      makeTable(TableIdentifier(tableName, Option(dbName)))
-    } else {
-      throw QueryCompilationErrors.tableOrViewNotFoundInDatabaseError(tableName, dbName)
-    }
+    // For backward compatibility (Spark 3.3 and prior), here we always look up the table from the
+    // Hive Metastore.
+    makeTable(Seq(CatalogManager.SESSION_CATALOG_NAME, dbName, tableName))
   }
 
   /**
    * Gets the function with the specified name. This function can be a temporary function or a
    * function. This throws an `AnalysisException` when no `Function` can be found.
    */
   override def getFunction(functionName: String): Function = {
-    // calling `sqlParser.parseFunctionIdentifier` to parse functionName. If it contains only
-    // function name and optionally contains a database name(thus a FunctionIdentifier), then
-    // we look up the function in sessionCatalog.
-    // Otherwise we try `sqlParser.parseMultipartIdentifier` to have a sequence of string as
-    // the qualified identifier and resolve the function through SQL analyzer.
-    try {
-      val ident = sparkSession.sessionState.sqlParser.parseFunctionIdentifier(functionName)
-      getFunction(ident.database.orNull, ident.funcName)
-    } catch {
-      case e: org.apache.spark.sql.catalyst.parser.ParseException =>
-        val ident = sparkSession.sessionState.sqlParser.parseMultipartIdentifier(functionName)
-        makeFunction(ident)
+    val parsed = sparkSession.sessionState.sqlParser.parseMultipartIdentifier(functionName)
+    // For backward compatibility (Spark 3.3 and prior), we should check if the function exists in
+    // the Hive Metastore first.
+    val nameParts = if (parsed.length <= 2 &&
+      !sessionCatalog.isTemporaryFunction(parsed.asFunctionIdentifier) &&
+      sessionCatalog.functionExists(parsed.asFunctionIdentifier)) {
+      CatalogManager.SESSION_CATALOG_NAME +: parsed
+    } else {
+      parsed
     }
+    makeFunction(nameParts)
   }
 
   /**
    * Gets the function with the specified name. This returns `None` when no `Function` can be
    * found.
    */
   override def getFunction(dbName: String, functionName: String): Function = {
-    makeFunction(FunctionIdentifier(functionName, Option(dbName)))
+    // For backward compatibility (Spark 3.3 and prior), here we always look up the function from
+    // the Hive Metastore.
+    makeFunction(Seq(CatalogManager.SESSION_CATALOG_NAME, dbName, functionName))
```

Comment:
Does this mean the previous code `makeFunction(FunctionIdentifier(functionName, Option(dbName)))` has a backward compatibility bug before?

Reply:
This really depends on if SQL analyzer respect current catalog when resolving Unresolvedfunc.

Reply:
The previous code is good as it always goes through the v1 code path. The new code appends the catalog name `spark_catalog` and goes through v2 code path, which is good as well.